### PR TITLE
Remove type and replace it with scheduler methods

### LINKED ISSUE #83514: podConditionUpdater and podPreemptor consist of a clientset only
After Moving podConditionUpdater and podPreemptor types to the scheduler, both of them only consists of `clientSet` which is already a part of the `scheduler` struct. https://github.com/kubernetes/kubernetes/pull/83342#discussion_r330850183 suggested that we remove them and have their methods as private methods of the scheduler.

They might as well be kept as they provide more context for the associated methods (i.e `setNominatedNodeName` during preemption, might be easier to understand the method usage when it's bound to `podPreemptor` type).

This issue is to further discuss whether or not we will remove those types.

cc @alculquicondor @ahg-g 

/kind cleanup
/sig scheduling

### PR DESCRIPTION
**Pull request by zouyee**
Signed-off-by: Zou Nengren <zouyee1989@gmail.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Remove type and replace it with scheduler methods which could simplify implementation logic and enhance code maintainability and readability.

**Which issue(s) this PR fixes**:

Fixes #83514

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -761,43 +737,62 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
...
 }
 
-type podConditionUpdaterImpl struct {
-	Client clientset.Interface
-}
-
-func (p *podConditionUpdaterImpl) update(pod *v1.Pod, condition *v1.PodCondition) error {
+// update updates the condition of a pod based on the passed PodCondition
+// used only in case of scheduling errors. If we succeed with scheduling,
+// PodScheduled condition will be updated in apiserver in /bind
```
Comment (denkensk) [inline pkg/scheduler/scheduler.go]: /bind --> bind

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -761,43 +737,62 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
...
+// handler so that binding and setting PodCondition it is atomic.
+func (sched *Scheduler) update(pod *v1.Pod, condition *v1.PodCondition) error {
```
Comment (denkensk) [inline pkg/scheduler/scheduler.go]: update->updatePodCondition ?

**Comment by ahmad-diaa** [comment]
Could you please reference the following issue in the description #83514, so that it gets closed when the PR is merged.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -761,43 +737,62 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
...
+	// This should only happen in unit tests
```
Comment (ahmad-diaa) [inline pkg/scheduler/scheduler.go]: I do not know if it's a good option to have code just for the tests within the method implementation. Can it be handled from the tests itself?

Comment (ingvagabund) [reply pkg/scheduler/scheduler.go]: @zouyee you can mock the kube client with `k8s.io/client-go/kubernetes/fake.NewSimpleClientset()` and inject predefined objects

File: pkg/scheduler/scheduler_test.go [hunk 4]
```diff
@@ -288,7 +264,7 @@ func TestScheduler(t *testing.T) {
 						return item.injectBindError
 					}}
 				},
-				podConditionUpdater: fakePodConditionUpdater{},
+				Client: nil,
```
Comment (ingvagabund) [inline pkg/scheduler/scheduler_test.go]: `TestScheduler` already defines `client := clientsetfake.NewSimpleClientset(&testNode)`. Depending on the execution path of individual tests under `TestScheduler`, you might need to pass pod object along side `testNode` into `NewSimpleClientset`.

Comment (zouyee) [reply pkg/scheduler/scheduler_test.go]: Done

**Comment by zouyee** [comment]
@ingvagabund PTAL

**Comment by zouyee** [comment]
@ahg-g PTAL

File: pkg/scheduler/scheduler.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/scheduler.go]: +1 to changing the name to updatePodCondition

File: pkg/scheduler/scheduler.go [hunk 5]
```diff
@@ -764,43 +740,41 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
...
-
-func (p *podPreemptorImpl) getUpdatedPod(pod *v1.Pod) (*v1.Pod, error) {
-	return p.Client.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+func (sched *Scheduler) getUpdatedPod(pod *v1.Pod) (*v1.Pod, error) {
+	return sched.Client.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
 }
 
-func (p *podPreemptorImpl) deletePod(pod *v1.Pod) error {
-	return p.Client.CoreV1().Pods(pod.Namespace).Delete(pod.Name, &metav1.DeleteOptions{})
+// deletePod delete a pod
```
Comment (ahg-g) [inline pkg/scheduler/scheduler.go]: please end the comment with a period (".")

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -85,13 +68,7 @@ type Scheduler struct {
...
 	Algorithm core.ScheduleAlgorithm
 	GetBinder func(pod *v1.Pod) Binder
-	// PodConditionUpdater is used only in case of scheduling errors. If we succeed
-	// with scheduling, PodScheduled condition will be updated in apiserver in /bind
-	// handler so that binding and setting PodCondition it is atomic.
-	podConditionUpdater podConditionUpdater
-	// PodPreemptor is used to evict pods and update 'NominatedNode' field of
-	// the preemptor pod.
-	podPreemptor podPreemptor
+	Client    clientset.Interface
```
Comment (ahg-g) [inline pkg/scheduler/scheduler.go]: make it local: s/Client/client

**Review by ahg-g — COMMENTED** [review]
sorry I missed this one, few nits and please rebase.

**Comment by ahg-g** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

Why do you keep getting this? Have you signed the CLA?

**Comment by zouyee** [comment]
> > Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> > **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> > It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> > 
> > * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> > * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> > * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> > * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)
> 
> Why do you keep getting this? Have you signed the CLA?

It's my mistake that forgot to set up email configuration when submitting code using my friend's computer.

File: pkg/scheduler/scheduler.go [hunk 7]
```diff
@@ -408,7 +383,7 @@ func (sched *Scheduler) recordSchedulingFailure(podInfo *framework.PodInfo, err
 // If it succeeds, it adds the name of the node where preemption has happened to the pod spec.
 // It returns the node name and an error if any.
 func (sched *Scheduler) preempt(ctx context.Context, state *framework.CycleState, fwk framework.Framework, preemptor *v1.Pod, scheduleErr error) (string, error) {
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go]: It looks like we don't have any unit tests covering sched.preempt?

Could you add some? We have some in `test/integration/preemption_test.go`, but maybe some of those should be unit instead.

File: pkg/scheduler/scheduler.go [hunk 8]
```diff
@@ -57,23 +57,6 @@ const (
 	pluginMetricsSamplePercent = 10
 )
 
-// podConditionUpdater updates the condition of a pod based on the passed
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go]: I'm kind of having second thoughts about this.

Maybe we should still get rid of the interfaces (and fakes), but preserve the implementation in a (single) separate struct called podClient.

Thoughts @ahg-g ?

File: pkg/scheduler/scheduler.go [hunk 9]
```diff
@@ -332,8 +309,6 @@ func New(client clientset.Interface,
 	sched.Recorder = recorder
 	sched.DisablePreemption = options.disablePreemption
 	sched.StopEverything = stopEverything
-	sched.podConditionUpdater = &podConditionUpdaterImpl{client}
-	sched.podPreemptor = &podPreemptorImpl{client}
 	sched.scheduledPodsHasSynced = podInformer.Informer().HasSynced
 	sched.client = client
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go]: I was actually getting rid of this in my next PR. See TODO in line 99

File: pkg/scheduler/scheduler_test.go [hunk 10]
```diff
@@ -276,10 +252,10 @@ func TestScheduler(t *testing.T) {
 			var gotBinding *v1.Binding
 			sCache := &fakecache.Cache{
 				ForgetFunc: func(pod *v1.Pod) {
-					gotForgetPod = pod
+					gotForgetPod = pod.DeepCopy()
```
Comment (haosdent) [inline pkg/scheduler/scheduler_test.go]: Is `DeepCopy` required?

**Pull request closed without merging**